File: src/main/java/network/matic/dagger/MqttRegex.java
package network.matic.dagger;

import network.matic.dagger.exceptions.DaggerException;

import java.util.regex.Pattern;

enum TokenType {
    SINGLE, MULTI, RAW;
}

class Token {
    private TokenType type;
    private String name;
    private String piece;
    private String last;

    public Token(TokenType type, String name, String piece, String last) {
        this.type = type;
        this.name = name;
        this.piece = piece;
        this.last = last;
    }

    public TokenType getType() {
        return type;
    }

    public String getName() {
        return name;
    }

    public String getPiece() {
        return piece;
    }

    public String getLast() {
        return last;
    }
}

public class MqttRegex {
    private String topic;
    private String rawTopic;
    private Pattern regexp;

    public MqttRegex(String t) throws DaggerException {
        String topic = t.toLowerCase();
        String[] tokens = MqttRegex.tokanize(topic);

        this.topic = String.join("/", tokens);
        this.rawTopic = topic;

        Token[] tokenObjects = new Token[tokens.length];
        for (int index = 0; index < tokens.length; index++) {
            tokenObjects[index] = MqttRegex.processToken(tokens[index], index, tokens);
        }

        this.regexp = MqttRegex.makeRegex(tokenObjects);
    }

    public boolean matches(String rawTopic) {
        String topic = rawTopic.toLowerCase();
        return this.regexp.matcher(topic).matches();
    }

    public String getTopic() {
        return topic;
    }

    public String getRawTopic() {
        return rawTopic;
    }

    public Pattern getRegexp() {
        return regexp;
    }

    public static String[] tokanize(String topic) {
        topic = topic.toLowerCase();
        String[] tokens = topic.split("/");
        if (tokens.length >= 4 && tokens[0].contains(":log")) {
            for (int i = 4; i < tokens.length; i++) {
                if (!tokens[i].equals("+") && !tokens[i].equals("#")) {
                    tokens[i] = Numeric.toStringPadded(tokens[i], 64);
                }
            }
        }

        return tokens;
    }

    public static Pattern makeRegex(Token[] tokens) {
        Token lastToken = tokens[tokens.length - 1];
        String[] regexTokens = new String[tokens.length];

        String[] result = new String[tokens.length];
        for (int index = 0; index < tokens.length; index++) {
            Token token = tokens[index];
            boolean isLast = index == tokens.length - 1;
            boolean beforeMulti = index == tokens.length - 2 && lastToken.getType() == TokenType.MULTI;
            result[index] = isLast || beforeMulti ? token.getLast() : token.getPiece();
        }

        return Pattern.compile(String.format("^%s$", String.join("", result)));
    }

    public static Token processToken(String token, int index, String[] tokens) throws DaggerException {
        boolean last = index == tokens.length - 1;

        if (token == null || "".equals(token.trim())) {
            throw new DaggerException("Topic must not be empty in pattern path.");
        }

        String cleanToken = token.trim();
        if (cleanToken.charAt(0) == '+') {
            return new Token(TokenType.SINGLE, "", "([^/#+]+/)", "([^/#+]+/?)");
        } else if (cleanToken.charAt(0) == '#') {
            if (!last) {
                throw new DaggerException("# wildcard must be at the end of the pattern");
            }
            return new Token(TokenType.MULTI, "#", "((?:[^/#+]+/)*)", "((?:[^/#+]+/?)*)");
        }

        return new Token(TokenType.RAW, cleanToken, String.format("%s/", cleanToken), String.format("%s/?", cleanToken));
    }
}

File: src/main/java/network/matic/dagger/MqttRegex.kt
package network.matic.dagger

import network.matic.dagger.exceptions.DaggerException
import network.matic.dagger.utility.Numeric
import java.util.regex.Pattern

open class MqttRegex(internal val t: String) {

    internal var topic: String
    internal val rawTopic: String
    internal val regexp: Pattern

    init {
        val topic = t.toLowerCase()
        val tokens: Array<String> = tokanize(topic)
        this.topic = java.lang.String.join("/", *tokens)
        rawTopic = topic
        val tokenObjects = arrayOfNulls<Token>(tokens.size)
        for (index in tokens.indices) {
            tokenObjects[index] = processToken(tokens[index], index, tokens)
        }
        regexp = makeRegex(tokenObjects)
    }

    fun matches(rawTopic: String): Boolean {
        val topic = rawTopic.toLowerCase()
        return regexp.matcher(topic).matches()
    }

    companion object {

        fun tokanize(topic: String): Array<String> {
            var topic = topic
            topic = topic.toLowerCase()
            val tokens = topic.split("/").toTypedArray()
            if (tokens.size >= 4 && tokens[0].contains(":log")) {
                for (i in 4 until tokens.size) {
                    if (tokens[i] != "+" && tokens[i] != "#") {
                        tokens[i] = Numeric.toStringPadded(tokens[i], 64)
                    }
                }
            }
            return tokens
        }

        fun makeRegex(tokens: Array<Token?>): Pattern {
            val lastToken = tokens[tokens.size - 1]
            val result = arrayOfNulls<String>(tokens.size)
            for (index in tokens.indices) {
                val token = tokens[index]
                val isLast = index == tokens.size - 1
                val beforeMulti = index == tokens.size - 2 && lastToken?.type === EnumHolder.TokenType.MULTI
                result[index] = if (isLast || beforeMulti) token?.last else token?.piece
            }
            return Pattern.compile(String.format("^%s$", java.lang.String.join("", *result)))
        }

        @Throws(DaggerException::class)
        fun processToken(token: String?, index: Int, tokens: Array<String>): Token {
            val last = index == tokens.size - 1
            if (token == null || "" == token.trim { it <= ' ' }) {
                throw DaggerException("Topic must not be empty in pattern path.")
            }
            val cleanToken = token.trim { it <= ' ' }
            if (cleanToken[0] == '+') {
                return Token(EnumHolder.TokenType.SINGLE, "", "([^/#+]+/)", "([^/#+]+/?)")
            } else if (cleanToken[0] == '#') {
                if (!last) {
                    throw DaggerException("# wildcard must be at the end of the pattern")
                }
                return Token(EnumHolder.TokenType.MULTI, "#", "((?:[^/#+]+/)*)", "((?:[^/#+]+/?)*)")
            }
            return Token(EnumHolder.TokenType.RAW, cleanToken, String.format("%s/", cleanToken), String.format("%s/?", cleanToken))
        }
    }
}
